apollo-server-core: Update schema reporting plugin to support gateways

sachindshinde opened the pull request:
To summarize, this PR:
- Updates Apollo Server to support a new `schemaDidLoadOrUpdate` event hook.
- Updates the Apollo Server schema reporting plugin to work when the server is configured with a gateway.

Note that this PR is dependent on apollographql/federation#738 :
- This PR has FIXMEs that require knowing the release version of gateway that lands the other PR.
- To use schema reporting with gateways, users must use a gateway with at least the above release version.

This PR specifically:
- Changes `GatewayInterface`to have a new method `onSchemaLoadOrUpdate`, for registering listeners that are notified on both schema load and schema update, and for providing the core schema in addition to the API schema.
  - This method is implemented in apollographql/federation#738 .
  - The `onSchemaChange` method has been marked deprecated and changed to optional, to support its eventual removal from the gateway API.
- Creates a new TS class `SchemaManager` for consolidating schema-tracking logic and abstracting away the gateway. This class specifically:
  - Acts as a source-of-truth for the schema and schema-derived data, updating it as needed.
  - Supports registering listeners for (and notifying listeners of) schema load/update events.
  - Is `async`-safe, i.e. it ensures linearizability for the TS class's public methods (provided `start()` is called and completes before `stop()`).
- Adds a new `schemaDidLoadOrUpdate` event hook so that plugins can listen to schema load/update events.
  - `ApolloServerBase._start()` registers such event hooks as schema load/update event listeners for the `SchemaManager`.
- Updates `ApolloServerPluginSchemaReporting` to use the new plugin event hook.
  - The hook, when called, creates a new `SchemaReporter` instance and starts it after stopping the old `SchemaReporter` instance (if it exists).
- Removes some states and properties from `ServerState` that are no longer necessary due to `SchemaManager`.
  - The two `initialized` states have been consolidated, and the `invoking serverWillStart` state has been removed.
  - The property `schemaDerivedData` has effectively been replaced by `schemaManager`, while `loadedSchema` has been removed entirely.
- Updates a test that was relying on a race condition between `start()` and `stop()` to resolve in a particular way to instead manually wait on `start()` before calling `stop()`.
- Updates plugin docs, migration docs, and changelog.

glasser commented:
I haven't reviewed the code yet, because your excellent PR description made it easy to reason about the design before doing so.

I think this is a bit of a non-starter:

> Have an additional argument for specifying the gateway instance when using schema reporting in a federated graph.

While we can make ApolloServer pass gateway automatically when it creates the schema reporting plugin, I think requiring users to do it themselves when they are calling ApolloServerPluginSchemaReporting is too error-prone. If you leave it out then it just silently doesn't show updates? Maybe there's some way to log a warning or throw in that case but I'm not sure how.

I think something like this would be better:
>We could make a new hook so that plugins can get access to the gateway instance, although I was hesitant to modify the plugin API to that extent (this is also just a use case of one plugin, so might be a bit hard to figure out the general shape at this point).

However instead of it being "provide gateway to plugin", I'd rather just introduce a new `schemaDidChange` callback on `GraphQLServerListener`, and make apollo-server invoke it on schema change.

(I'd suggest that it takes an object with keys as an argument rather than two positional arguments, to be consistent with the rest of the plugin API. So it will be shaped differently from onSchemaChange, but that's OK?)

(I think it's fine to "miss" changes that happen before serverWillStart — the plugin will get whatever the schema is at the time of serverWillStart, and any afterwards, and it's OK if it misses some schemas that predate it running. As long as no change *after* serverWillStart is missed!)

What do you think? Let me know if you'd like me to review the code now or wait for an update based on this (or if you think the current implementation is better of course).

glasser commented on packages/apollo-server-core/src/plugin/schemaReporting/schemaReporter.ts:
```diff
@@ -117,11 +117,13 @@ export class SchemaReporter {
       const result = await this.reportServerInfo(sendNextWithExecutableSchema);
       switch (result.kind) {
```
If you'd like to factor this one out into its own PR we could get it merged fast!

sachindshinde replied:
Sounds good, moved this change to #5222 🙂

sachindshinde commented:
@glasser 
> While we can make ApolloServer pass gateway automatically when it creates the schema reporting plugin, I think requiring users to do it themselves when they are calling ApolloServerPluginSchemaReporting is too error-prone. If you leave it out then it just silently doesn't show updates? Maybe there's some way to log a warning or throw in that case but I'm not sure how.

Before this PR, the code in `serverWillStart` in `packages/apollo-server-core/src/plugin/schemaReporting/index.ts` would check whether the schema is federated via `schemaIsFederated()`, and throw if it was. This PR keeps that check, but only throws if no `gateway` instance was provided. (Sorry for any confusion here, I think it phrased it badly in the PR description, so I've updated it to clarify here.)

>However instead of it being "provide gateway to plugin", I'd rather just introduce a new `schemaDidChange` callback on `GraphQLServerListener`, and make apollo-server invoke it on schema change.

I'd be down for introducing a new `schemaDidChange` callback for plugins to specify on `GraphQLServerListener`.

>I'd suggest that it takes an object with keys as an argument rather than two positional arguments, to be consistent with the rest of the plugin API. So it will be shaped differently from onSchemaChange, but that's OK?

Yep, that sounds fine to me.

>(I think it's fine to "miss" changes that happen before serverWillStart — the plugin will get whatever the schema is at the time of serverWillStart, and any afterwards, and it's OK if it misses some schemas that predate it running. As long as no change after serverWillStart is missed!)

Agreed here, should be fine to miss updates prior to `serverWillStart` running.

>What do you think? Let me know if you'd like me to review the code now or wait for an update based on this (or if you think the current implementation is better of course).

I've updated this PR in the last six commits to introduce a new `schemaDidChange` hook/callback, which takes in a `GraphQLSchemaContext` object with an API schema and an optional core schema SDL.

The gist of the changes is that:
- `ApolloServerBase._start()` adds any `schemaDidChange` plugin callbacks as listeners to gateway schema changes via `onSchemaChange`.
  - There's a bit of complexity here due to an edge-case where the gateway updates while `serverWillStart` is executing, in which case `schemaDidChange` may miss updates. To get around this, I added a listener to gateway schema changes to record the latest gateway schema, and manually call `schemaDidChange` with that schema before adding `schemaDidChange` as a listener.
    - This could be cleaned up slightly if we start tracking `coreSchemaSdl` in `SchemaDerivedData`, though I'm not sure if we want to start doing that.
  - There's a bit of thorniness in that errors thrown by `schemaDidChange` get handled by whatever notifies `onSchemaChange` listeners in gateway, although I'm not sure if there's a nice way around this here.
- In `ApolloServerPluginSchemaReporting`, we no longer take in a gateway. Most of the code of `serverWillStart` shifts into the `schemaDidChange` callback.
  - I've changed the second positional argument of `SchemaChangeCallback` to be optional as you've suggested [here](https://github.com/apollographql/federation/pull/738#discussion_r631435298). This would mean that gateways being the wrong version results in a runtime error instead of a compile-time error, but IMO that's fine. (We can give better error messaging for runtime errors, and we already have a lot of runtime checks being done in `ApolloServerBase._start()` regardless so I don't think it'll be surprising.)
  - Note the code currently assumes `onSchemaChange` callbacks are called on the initial schema load. If that changes, I'll need to modify `ApolloServerBase._start()` a bit to use the output of `load()` when the `onSchemaChange` callback hasn't been called.
  - Note the code currently assumes `schemaDidChange` callbacks would be similarly called on initial schema load. If we want different semantics there, I'll need to modify `serverWillStart`'s typings to pass an (optional) core schema SDL and the `serverWillStart` callback in `ApolloServerPluginSchemaReporting` to create a `SchemaReporter`.

Let me know if this looks fine, if so I'll modify the PR description accordingly for the new approach 🙂

glasser commented on packages/apollo-server-types/src/index.ts:
```diff
@@ -84,6 +84,11 @@ export interface ApolloConfig {
 
+export interface GraphQLSchemaContext {
+  apiSchema: GraphQLSchema,
+  coreSchemaSdl?: string,
```
Can you leave a comment noting that this will only be left out with old versions of Gateway and that we should consider making this required for AS3 (which can definitely have a minimum version of Gateway)? (Though my long comment suggests we can probably make this required now.)

sachindshinde replied:
This argument may still be gone I think, specifically if `schemaDidLoadOrUpdate` is used by a plugin when a gateway isn't provided (which is the case for non-federated schema reporting).

glasser commented on packages/apollo-server-plugin-base/src/index.ts:
```diff
@@ -65,6 +67,9 @@ export interface ApolloServerPlugin<
 export interface GraphQLServerListener {
+  schemaDidChange?(
+    schemaContext: GraphQLSchemaContext,
+  ): void;
```
Would it make the correctness a lot harder to make this method async (ideally just plain Promise<void> async as opposed to the weirder ValueOrPromise)? Making all plugin methods async is in the air for AS3 (#4999). I can imagine it might make the logic harder to implement correctly.

sachindshinde replied:
(Paraphrasing from a Slack convo we had with Trevor) We've decided to keep this sync, as there's concerns about listeners making expensive async calls stalling execution. If users want to do async things in response to a schema load/update, they can queue the changes and manage async queue consumption themselves.

glasser commented on packages/apollo-server-core/src/ApolloServer.ts:
```diff
@@ -563,15 +583,49 @@ export class ApolloServerBase {
+            // Note that it's important for there to be no await between this schemaDidChange()
+            // invocation and the registration of the new schema change callback, otherwise plugins
+            // could miss schema changes.
+            schemaDidChange({
```
I continue to not like the idea of calling schemaDidChange redundantly with the serverDidStart with the same schema. This ties in to what we are discussing with @trevor-scheer over at https://github.com/apollographql/federation/pull/738#discussion_r634948971
I'd really prefer "did change" to actually mean "did change" and not "here is the same information we already gave you again".

If we're concerned about the serverWillStart overlapping with update thing, that makes sense: but I think we should make the change I want in 738 and then instead of asserting that `latestApiSchema` must be set, use whether or not it is set to call schemaDidChange only in the case of this overlap.

Here's a thought: what if in addition to `gateway.onSchemaChange`, we also allowed you to pass an onSchemaChange handler to `gateway.load`, with the semantics that this handler would be installed immediately after the initial load and it's gateway's job to make sure no changes are missed? Would that let us simplify things?

Or hmm.

Hmm.

OK, having looked at the whole PR, I do see that there is something simplifying about having a callback that is called every time the schema is updated, including the first time. Maybe my objection is just to the "changed" terminology. Can we come up with a more precise term that doesn't imply a change from a previous value? `schemaDidLoad` probably implies only the first time.

If we're doing that, we could consider changing #738 so that instead of adding a second argument to the callback, we just add a new `onSchemaWhatever` whose callback takes an object with keys instead of separate arguments.

And ooh! I think this will let us make the ApolloServer schemaDidWhatever take coreSupergraphSdl as a non-`?` argument. The idea is that we can move the runtime check into the schema loading code from the schema reporter. We can basically do:

- If any plugin registered schemaDidWhatever, and there is a gateway, then `gateway.onSchemaWhatever` must exist or it's an immediate startup error.  (ie, it's an error to combine schema reporting with an old gateway, which shouldn't break any existing users since that's already an error.)
- However, if no plugin registered schemaDidWhatever, then we can fall back to `gateway.onSchemaChange` if onSchemaWhatever doesn't exist.

sachindshinde replied:
I've changed the name to `schemaDidLoadOrUpdate` and changed apollographql/federation#738 to have a new `onSchemaLoadOrUpdate` method that takes an object with keys/properties.

>And ooh! I think this will let us make the ApolloServer schemaDidWhatever take coreSupergraphSdl as a non-`?` argument.

I think that argument still ends up needing to be `?` to be for the non-federated schema reporting case.

>The idea is that we can move the runtime check into the schema loading code from the schema reporter. We can basically do:
>- If any plugin registered schemaDidWhatever, and there is a gateway, then `gateway.onSchemaWhatever` must exist or it's an immediate startup error. (ie, it's an error to combine schema reporting with an old gateway, which shouldn't break any existing users since that's already an error.)

Sounds good, I've added a check for this near the end of `_start()`. Note that I'm beginning to think no error is currently thrown when trying to use the schema reporting plugin with a gateway (or at least, I don't see any code that throws when that's the case). However, our autoreg pipeline silently throws away reports from gateways.

>- However, if no plugin registered schemaDidWhatever, then we can fall back to `gateway.onSchemaChange` if onSchemaWhatever doesn't exist.

We have to register gateway listeners early in `_start()` before we ask plugins whether they provide `schemaDidLoadOrUpdate`, so the new code just uses `schemaDidLoadOrUpdate` if available or falls back to `onSchemaChange` (regardless of plugin registration).

glasser commented on packages/apollo-server-core/src/plugin/schemaReporting/index.ts:
```diff
@@ -98,21 +99,19 @@ export function ApolloServerPluginSchemaReporting(
-
+      const isFederatedSchemaReporting = schemaIsFederated(schema);
       if (overrideReportedSchema !== undefined) {
+        if (isFederatedSchemaReporting) {
```
I'm pretty sure schemaIsFederated is looking for the case that this is a *subgraph* schema, so neither of the places you use this bool seem right.

Are we planning to do anything about schema reporting from subgraphs?

sachindshinde replied:
You're right, good catch! I've reverted this part of the code, so that we use `schemaIsFederated` only to throw for schemas that look like implementing service schemas.

>Are we planning to do anything about schema reporting from subgraphs?

I'm not in the loop enough about future plans there. For now I think we intend to keep the status quo, but we might start having reporting there in the future.

glasser commented on packages/apollo-server-core/src/ApolloServer.ts:
```diff
@@ -563,15 +583,49 @@ export class ApolloServerBase {
+            });
+            unsubscribe = maybeGateway.onSchemaChange(
+              (apiSchema, coreSchemaSdl) => {
+                schemaDidChange({ apiSchema, coreSchemaSdl });
```
It seems like this needs to be better synchronized with the logic in the existing onSchemaChange. eg, we shouldn't call it if we don't actually update the schema because we're stopping or something, and we shouldn't call it until after the schema is actually updated (ie, schemaDerivedData is successfully changed).  Probably this means we need to save the handlers somewhere and call them from the existing onSchemaChange.

sachindshinde replied:
Agree here, I've changed `_start()` to add listeners to a set and the gateway listener created in `startGatewayAndLoadSchema()` to notify listeners in the set after the schema derived data is updated.

sachindshinde commented:
@glasser @trevor-scheer 
Update here, I've changed the PR to be more in line with what's mentioned above and in Slack. Specifically:
- The new hook has been renamed to `didSchemaLoadOrUpdate`, to better indicate that is called on both initial schema load and on schema updates.
- In `startGatewayAndLoadSchema()`, we use `onSchemaLoadOrUpdate` instead of `onSchemaChange` if available.
  - Previously, the gateway listener would only store the schema (and derived data) if the phase was `started`, but this leads to a potential issue where:
    - `gateway.load()` runs and returns the initial schema.
    - Somewhere in between then and the setting of the phase to `started`, an `await` occurs (e.g. executing hooks), halting execution.
    - Gateway's schema updates and notifies listeners, but our gateway listeners ignore the change because the state isn't `started`.
    - Execution resumes from the `await`, and the rest of `_start()` continues assuming the schema is that initial schema.
    - AS continues to believe the schema is the initial schema until the next gateway schema update, which may be arbitrarily far into the future. (Effectively, an update has been missed.)
  - This PR changes the listener to also store the schema if the phase is `starting` or `invoking serverWillStart`.
- We don't add multiple listeners to gateway, and instead add `didSchemaLoadOrUpdate` listeners to a set.
  - If a gateway is provided, and any plugin provides a `didSchemaLoadOrUpdate` listener, we throw in `_start()`, saying gateway must be updated.
  - Before adding a listener to the set, we call the listener with the latest schema in `this.state.schemaDerivedData`.
- The `onSchemaLoadOrUpdate`/`onSchemaChange` listener added in `startGatewayAndLoadSchema()` then forwards to the `didSchemaLoadOrUpdate` listeners in the set after storing the schema.

Once the code's approved, I'll update the PR description and the `CHANGELOG.md`.

sachindshinde commented:
Update here on testing/dogfooding.

Gateway schema reporting is currently running in Studio staging, and appears to be working fine for both the old GCS pipeline and the new configdelivery pipeline.

I've locally tested gateways with a service list (to test the `loadStatic()` path) and plain no-gateway/monolith apollo server, and schema reporting is working there as well.

glasser commented on packages/apollo-server-core/src/ApolloServer.ts:
```diff
@@ -512,27 +522,44 @@ export class ApolloServerBase {
         phase: 'invoking serverWillStart',
         barrier,
-        schemaDerivedData,
+        // We can't just use the variable schemaDerivedData here, as any
```
I mean, this complexity is exactly why I wanted to differentiate between initial load and later updates, and why my instinct was to move away from "use the same callback for initial load and later update". I feel like the old logic of "it's the job of gateway.load to ensure that at the moment it resolves its promise, that's the current schema, and as long as you don't yield between `await load()` and updating the state and starting to listen for changes, it works out.

I don't want to jerk you back and forth here... but I thought the point of doing onSchemaLoadOrUpdate was to make this stuff less complex, not more complex.  This is really hard to follow right now.

sachindshinde replied:
>I mean, this complexity is exactly why I wanted to differentiate between initial load and later updates, and why my instinct was to move away from "use the same callback for initial load and later update". I feel like the old logic of "it's the job of gateway.load to ensure that at the moment it resolves its promise, that's the current schema, and as long as you don't yield between `await load()` and updating the state and starting to listen for changes, it works out.

I'm not sure separating initial load and later updates would help for this specific problem. Specifically, the following can happen:
- While executing `_start()`, `await gateway.load()` runs, which executes `gateway.load()`.
- At some point during the execution of `gateway.load()`, it returns a promise that resolves to the initial schema, and execution suspends while waiting on that promise. Let's suppose the promise resolves at time A.
- At some point later (time B), `_start()` resumes execution (now that the promise has resolved, execution is unblocked).
- Between A and B, it's possible for other code to execute, including the code that updates the schema.

Basically it's hard to keep the initial schema non-stale when it's dynamic.

>I don't want to jerk you back and forth here... but I thought the point of doing onSchemaLoadOrUpdate was to make this stuff less complex, not more complex. This is really hard to follow right now.

FWIW, I think once AS3 lands and we can require gateways to define `onSchemaLoadOrUpdate`, this changes to a:
```typescript
this.state = {
  // ...
  schemaDerivedData: this.state.schemaDerivedData,
}
```
like the other changes to `this.state`, and `this.state.schemaDerivedData` effectively becomes the source of truth.

glasser replied:
Maybe we could do a (noninvasive) "polyfill"? Like we can define a wrapper "gateway" which, if gateway.onSchemaLoadOrUpdate doesn't exist, we wrap around the gateway. It passes through methods other than onSchemaLoadOrUpdate and load, and handles onSchemaLoadOrUpdate by delegating to onSchemaChange and also keeping it around to call after load?  So we could put all the back compat logic in this one wrapper class, and keep the main startup logic clean?

sachindshinde replied:
@glasser I ended up going with something a bit more general than a polyfill here, it's at `packages/apollo-server-core/src/utils/schemaManager.ts` (it wraps gateway and provides `onSchemaLoadOrUpdate`, but it also wraps schemas in the non-gateway case and is responsible for updating schema-derived data).

glasser commented on packages/apollo-server-core/src/ApolloServer.ts:
```diff
@@ -124,7 +127,11 @@ type SchemaDerivedData = {
+  | {
+      phase: 'starting';
+      barrier: Resolvable<void>;
+      schemaDerivedData?: SchemaDerivedData;
```
To me, the difference between the "starting" and "invoking serverWillStart" states is that the latter has the (initial) schema set up and derived, and the former doesn't. I'm not sure what the point is of having them as separate states otherwise.  I guess you could merge them.  But really I think it shouldn't be rocket science for us to get it working with distinct states that represent distinct moments in moving forward?

sachindshinde replied:
I ended up eliding out the `starting` since it was never distinguished from `invoking serverWillStart` in the updated PR.

glasser commented on packages/apollo-server-core/src/ApolloServer.ts:
```diff
@@ -564,16 +591,59 @@ export class ApolloServerBase {
+            // It is crucial there not be any await statements between this
+            // read and the addition of a listener to the listener set, or else
+            // schema updates could be missed by listeners.
+            const latestSchemaDerivedData = this.state.schemaDerivedData;
```
This is all far too complicated for me to understand or have any hope of successfully maintaining. Are there other approaches we can use to deal with the possibility of receiving schema change updates during startup?

For example, could we just buffer them?  If a schema change comes in during 'invoking serverWillStart' we save it to some field (and I think it's fine if we just have one field rather than an array — if two schema changes come in we can just ignore the non-last one), and after serverWillStart is done we can fire off any buffered schema?

ie, that would let us just consistently use `const schemaDerivedData` for all relevant calls during the course of `_start`. And instead of changing the hook from "ignore in non-started state" to "ignore in all but three states", it would be "buffer in the starting/serverWillStart states, call in started, ignore otherwise".

sachindshinde replied:
>This is all far too complicated for me to understand or have any hope of successfully maintaining. Are there other approaches we can use to deal with the possibility of receiving schema change updates during startup?

In the refactor, I've tried to address this by confining most of this code to non-async functions, which I feel makes it easier to reason about since such functions can't be suspended during execution.

>For example, could we just buffer them? If a schema change comes in during 'invoking serverWillStart' we save it to some field (and I think it's fine if we just have one field rather than an array — if two schema changes come in we can just ignore the non-last one), and after serverWillStart is done we can fire off any buffered schema?

I added some buffering in the `SchemaManager` class, there's more details about it in the `SchemaManager.onSchemaLoadOrUpdate` docstring.

sachindshinde commented:
@glasser Finally got some time to come back to this. I've wrapped all modes of operation (old gateway, new gateway, and non-gateway) into a class `SchemaManager` that:
- Provides a `onSchemaLoadOrUpdate()` method for listener registration that's relatively consistent across modes. (It also buffers the latest event for new listeners).
- Tracks all schema-derived data, updating it as needed via a provider callback passed in at construction.
- Confines most of the tricky code to non-async functions, which I've generally found makes reasoning about correctness a lot easier.

After moving `schemaDerivedData`, I was able to consolidate the two `initialized` states together, and consolidate the `starting` and `invoking serverWillStart` states together.

glasser commented:
Can you rebase onto `release-3.0`? Not planning any more AS2 releases.

sachindshinde commented:
@glasser Sounds good, I've rebased this branch on `release-3.0` and squashed a bunch of old commits away so the commit history should be cleaner.

glasser commented on packages/apollo-server-core/src/ApolloServer.ts:
```diff
@@ -419,6 +420,27 @@ export class ApolloServerBase {
+            try {
+              schemaManager.onSchemaLoadOrUpdate(schemaDidLoadOrUpdate);
+            } catch (_) {
+              // TODO: Once a gateway version providing the core schema to
```
If this is something that needs to be fixed before the PR can be merged, include "FIXME" so we don't forget.

sachindshinde replied:
Yep, this is something we should fix prior to merge (once we release gateway that has `onSchemaLoadOrUpdate()`).

I've also fixed a bug in the error message here (the name of the hook was wrong).

glasser commented on packages/apollo-server-core/src/ApolloServer.ts:
```diff
@@ -419,6 +420,27 @@ export class ApolloServerBase {
+          if (schemaDidLoadOrUpdate) {
+            try {
+              schemaManager.onSchemaLoadOrUpdate(schemaDidLoadOrUpdate);
+            } catch (_) {
```
I'm not sure I understand this try/catch.  schemaManager.onSchemaLoadOrUpdate can throw for multiple reasons, and one of those reasons already talks about gateway upgrades?

sachindshinde replied:
The idea here was that:
- The error message thrown by `SchemaManager.onSchemaLoadOrUpdate()` on old gateways doesn't actually mention the hook name `schemaDidLoadOrUpdate`, since `SchemaManager.onSchemaLoadOrUpdate()` may in the future be called in other locations. So the try/catch is at the callsite to change the error message to mention the hook name.
- As for the `catch (_)`, I believe `SchemaManager.onSchemaLoadOrUpdate()` right now will only throw for that one specific case. This is admittedly fragile though, so I've switched to using a specific error subclass, and translating the error message just for that subclass (rethrowing otherwise).

I'm also down for any other approaches here you have in mind.

glasser commented on packages/apollo-server-core/src/plugin/schemaReporting/index.ts:
```diff
@@ -152,33 +141,63 @@ export function ApolloServerPluginSchemaReporting(
+                [
+                  `The overrideReportedSchema option is incompatible with gateways`,
+                  `as the schema SDL is given directly by gateway. If you would`,
+                  `like to customize the schema SDL reported, please instead set`,
```
This doesn't just customize the reported SDL though right? It customizes the actual schema used?

sachindshinde replied:
Good catch! I've changed the code to accept `overrideReportedSchema` for gateways, interpreting it as the core schema (similar to usage reporting).

glasser commented on packages/apollo-server-core/src/plugin/schemaReporting/index.ts:
```diff
@@ -152,33 +141,63 @@ export function ApolloServerPluginSchemaReporting(
+              );
+            }
+            logger.info(
+              'Apollo schema reporting: schema to report has been overridden',
```
Will this print more than once ever?  No, because in the gateway case it's caught by the previous check?

sachindshinde replied:
It wouldn't before, but I've now removed the check, so I've accordingly added code to ensure the hook early returns after executing once in the case when the schema is overridden.

glasser commented on packages/apollo-server-core/src/ApolloServer.ts:
```diff
@@ -85,18 +80,16 @@ type SchemaDerivedData = {
-      schemaDerivedData: SchemaDerivedData;
+      schemaManager: SchemaManager;
     }
   | { phase: 'failed to start'; error: Error; loadedSchema: boolean }
```
While you're here you can remove `loadedSchema` which hasn't been read since https://github.com/apollographql/apollo-server/commit/49a71e4e01a8dbdb5c7151d9467ad0c164470cb6

sachindshinde replied:
Sounds good, I've removed `loadedSchema` from the server state.

glasser commented on packages/apollo-server-core/src/types.ts:
```diff
@@ -54,12 +55,28 @@ export type GraphQLServiceConfig = {
   load(options: { apollo: ApolloConfig }): Promise<GraphQLServiceConfig>;
+
+  /**
+   * @deprecated Use `onSchemaLoadOrUpdate` instead
```
Let's make this optional so we are eventually able to remove it?

sachindshinde replied:
I'd be down for this. It would cause TS failures for people compiling against `GatewayInterface`, but this should be few (if any) people, and it's arguably a decent way of notifying them that this method is disappearing soon (and that they should change their code accordingly).

glasser commented on packages/apollo-server-integration-testsuite/src/ApolloServer.ts:
```diff
@@ -2695,15 +2695,23 @@ export function testApolloServer<AS extends ApolloServerBase>(
-        await stopServer();
-        expect(unsubscribeSpy).toHaveBeenCalled();
-      });
+      // Serverless frameworks execute ApolloServer.start() in a dangling
```
Another alternative is to perform an action that requires startup to have finished, eg `await server.executeOperation({query: '{__typename}'})`

sachindshinde replied:
That sounds fine to me, I've re-enabled the test for serverless and added the above code to ensure it's started.

glasser commented on packages/apollo-server-plugin-base/src/index.ts:
```diff
@@ -49,13 +51,17 @@ export interface ApolloServerPlugin<
 }
 
 export interface GraphQLServerListener {
+  schemaDidLoadOrUpdate?(schemaContext: GraphQLSchemaContext): void;
```
In Apollo Server 3, all plugin callbacks are async other than willResolveField (I am paranoid about the performance impact of making that one async, though I did not actually benchmark).

Would it be possible to make this one async too, or would that introduce race conditions?

In any case, the docs (docs/source/integrations/plugins-event-reference.md and perhaps docs/source/integrations/plugins.md) should be updated with the new hook (possibly updating the parts that explain that all but one hook is async).

glasser replied:
It could be fair keep this one as synchronous to make SchemaManager simpler.

Or it could be async like everything else and just document that there's nothing stopping calls to them from overlapping.  That might be simpler to think about.  (and processSchemaLoadOrUpdateEvent wouldn't await, it would just `.catch` with the same logger we already have.)

In any case, docs need updating.

sachindshinde replied:
I'd say the main issues with async callbacks here are that they'd indeed increase the complexity of `SchemaManager`, and that it's not necessarily clear what the desired ordering semantics would be for the callbacks, e.g. with questions like:
- For a given schema load/update event, should its callbacks be executed concurrently or in some consistent sequential order?
- For a given schema load/update event, should the competion of its callbacks block subsequent schema updates?
  - If not, then for two schema load/update events, should all the callbacks of the prior event complete before all the callbacks of the subsequent event start?
    - If not that either, then for two schema load/update events, for each callback, should that callback complete for the prior event before the callback is started for the subsequent event?

I'm not necessarily sure what behavior users would want here, so I'd rather figure that out first and build it later.

(Still in the middle of updating the docs.)

glasser reviewed (changes requested):
Reviewed everything but schemaManager.ts so far.

glasser commented on packages/apollo-server-core/src/utils/schemaManager.ts:
```diff
@@ -0,0 +1,216 @@
+  /**
+   * Registers a listener for schema load/update events. Note that the latest
+   * event is buffered, i.e.
+   * - If registered before start(), then the callback will be first called in
```
This isn't actually possible in the way ApolloServer uses it, right? We could just make it throw?

sachindshinde replied:
Yep, `this.state` is private so users can't publicly call `onSchemaLoadOrUpdate()` before `start()`, and we don't call `onSchemaLoadOrUpdate()` before `start()` in our own code. I've changed the code in `onSchemaLoadOrUpdate()` to throw when called before `start()`.

glasser commented on packages/apollo-server-core/src/utils/schemaManager.ts:
```diff
@@ -0,0 +1,216 @@
+      !this.modeSpecificState.gateway.onSchemaLoadOrUpdate
+    ) {
+      throw new Error(
+        'Your gateway is too old to register a onSchemaLoadOrUpdate() listener',
```
It would be nice if we could get an `@apollo/gateway` version number to upgrade to in here, though it'll take some placeholdering until we get releases coordinated.

sachindshinde replied:
Yep, agreed it would be better to have a version number here, I've added a `FIXME` to this.

glasser commented on packages/apollo-server-core/src/utils/schemaManager.ts:
```diff
@@ -0,0 +1,216 @@
+        try {
+          callback(this.schemaContext);
+        } catch (e) {
+          this.logger.error(
```
This method is only called during `ApolloServer._start`, so if the callback throws (eg with "The overrideReportedSchema option is incompatible with gateways") it would be reasonable to make startup fail. Let's not get back into that weird world where errors during startup are just logged instead of causing startup to fail, even for something like these plugins. (Wrapping the error in something naming the plugin hook and rethrowing sounds reasonable though.)

Logging is reasonable for errors that occur *after* system startup here, but the whole point of `ApolloServer.start()` is that it's a thing that can easily make the server fail before listening if anything goes wrong.

sachindshinde replied:
Since it's only called in `_start()`, I'm down for having `onSchemaLoadOrUpdate()` catch and re-throw here. I've kept the logging as well, since it would appear JS `Error`s don't have a `cause` argument like Java `Throwable`s.

glasser replied:
I generally am not super into "log and rethrow" — if you're rethrowing then the caller should be able to catch and handle without there also being a log they have to suppress?

sachindshinde replied:
The main reason I'm logging there is in the event they want the stacktrace of the original error thrown by the plugin (which would disappear in the rethrow). But arguably they could add a catch/log/rethrow to their plugin code if they were interested in that, so I'm also down to remove the log.

glasser commented on packages/apollo-server-core/src/utils/schemaManager.ts:
```diff
@@ -0,0 +1,216 @@
+          this.logger.error(
+            [
+              "An error was thrown from an 'onSchemaLoadOrUpdate' listener:",
+              `${e?.message ?? e}`,
```
Doesn't this lose the stacktrace?

sachindshinde replied:
Good catch! I'm down to write it like
```typescript
this.logger.error(
  "An error was thrown from an 'onSchemaLoadOrUpdate' listener",
);
this.logger.error(e);
```
so we can keep the stacktrace (that seems to be the pattern at least for keeping stacktrace info, based on [this](https://github.com/apollographql/apollo-server/blob/main/packages/apollo-server-core/src/ApolloServer.ts#L256-L257)).

glasser reviewed (commented):
This overall looks great. Much easier for me to understand than last time.

sachindshinde commented on packages/apollo-server-core/src/ApolloServer.ts:
```diff
@@ -274,7 +268,15 @@ export class ApolloServerBase {
+        schemaManager: new SchemaManager({
+          gateway,
+          apolloConfig: this.apolloConfig,
+          schemaDerivedDataProvider: this.generateSchemaDerivedData.bind(this),
```
Sounds good, it's definitely clearer as `schemaDerivedDataProvider: (schema) => this.generateSchemaDerivedData(schema)` so I've switched it accordingly.

sachindshinde commented:
Responded to PR feedback, and updated `CHANGELOG.md` and the PR description.

glasser commented on packages/apollo-server-integration-testsuite/src/ApolloServer.ts:
```diff
@@ -2699,7 +2705,19 @@ export function testApolloServer<AS extends ApolloServerBase>(
+          // started in the case of serverless, we make a query against it,
+          // which forces us to wait until after start.
+          //
+          // This is also required because ApolloServer.stop() was not designed
```
Good point, not great that there's a lack of support for this (given that stop can get called by signal handlers). Want to add an XXX comment to ApolloServer.stop() noting that it might not behave well when called concurrently with start (or before it, even)?

sachindshinde replied:
Yep, I've added an XXX comment to `ApolloServerBase.stop()` to warn about the behavior.

glasser commented on packages/apollo-server-types/src/index.ts:
```diff
@@ -265,12 +270,14 @@ export interface CachePolicy extends CacheHint {
    * make the policy more restrictive or less restrictive.
    */
   replace(hint: CacheHint): void;
+
```
Curious where these extra newlines in some files came from, given that we do have prettier enforcement now...

sachindshinde replied:
I'm curious as well, my IntelliJ adds those newlines, but it looks CI prettier enforcement doesn't mind?

glasser replied:
Ah, interesting. Ideally it's nice not to put in unnecessary style changes (would be better to configure IntelliJ to use prettier instead of its own formatting) but not gonna sweat it for this PR.

sachindshinde replied:
What's odd is that I have IntelliJ configured to use prettier on format, but it's still adding the newlines (maybe it's a prettier version mismatch?)
<img width="996" alt="Screen Shot 2021-07-22 at 2 32 45 PM" src="https://user-images.githubusercontent.com/2695578/126690912-0abc69cd-2521-4c44-af89-7f2fef45fab6.png">

But yeah, I'm down to worry about that later.

glasser reviewed (changes requested):
OK this is almost perfect!

sachindshinde commented on docs/source/integrations/plugins-event-reference.md:
```diff
@@ -171,6 +171,35 @@ const server = new ApolloServer({
 
+### `schemaDidLoadOrUpdate`
+
+The `schemaDidLoadOrUpdate` event fires whenever Apollo Server initially loads the schema or updates the schema. A `schemaDidLoadOrUpdate` handler is given the new API schema and optionally the new core schema (if using a gateway). If you provide a gateway older than FIXME , attempting to register a `schemaDidLoadOrUpdate` handler will fail.
```
Sounds good, I've updated the phrasing as you've suggested.

sachindshinde commented on packages/apollo-server-core/src/ApolloServer.ts:
```diff
@@ -735,18 +741,6 @@ export class ApolloServerBase<
-          );
-        }
-      }
       if (!alreadyHavePlugin) {
```
Yep, that's equivalent and simpler. I've refactored the `if`/`else`s accordingly.

sachindshinde commented on CHANGELOG.md:
```diff
@@ -14,6 +14,8 @@ The version headers in this history reflect the versions of Apollo Server itself
 - `apollo-server-micro`: Like the other framework integrations, only serve landing pages from the GraphQL path (`/graphql` by default, configurable via the `path` option to `createHandler`). [PR #5516](https://github.com/apollographql/apollo-server/pull/5516)
 - `apollo-server-env`: Remove polyfills of `Object.values`, `Object.entries`, and `util.promisify` which were only required for Node 6 support. Remove `ValueOrPromise` and `WithRequired` TypeScript types that are also provided by `apollo-server-types`. [PR #5515](https://github.com/apollographql/apollo-server/pull/5515)
+- `apollo-server-core`/`apollo-server-plugin-base`: Add support for `schemaDidLoadOrUpdate` event hooks, to be specified by the `serverWillStart` event hook. Plugins listening for this event will receive the API schema (and core schema for gateways) when the server's schema is initially loaded and when the server's schema is updated. For more information about this plugin event, see [the plugin e ...
+- `apollo-server-core`: Add support for schema reporting when using Apollo Gateway. Note that your gateway must be at least version FIXME , or else attempts to enable schema reporting will fail. [PR #5187](https://github.com/apollographql/apollo-server/pull/5187)
```
Sounds good, I've added a bit about how Studio doesn't support gateway schema reporting yet and that they shouldn't use the feature (unless we tell them to).

glasser commented:
I pushed a small CHANGELOG change to move the updates to what is now the vNEXT section and to tweak the wording about Studio

glasser reviewed (approved):
Looks great. Let's make https://github.com/apollographql/federation/pull/738 mergeable before we merge this and merge them both when we're both free together.
